Support recognizing TypeScript files imported with `.js` suffix

In TypeScript, it is valid to import TypeScript files with `.js` extension. TypeScript will match the coresponding `.ts` files if they exist. This behaviour is commonly used in Node.js projects as [`.js` file extensions are required in Node.js ESM imports](https://nodejs.org/api/esm.html#esm_mandatory_file_extensions).

It would be great if ts-unused-exports also recognize variable imported with `.js` suffix.

Related: https://github.com/evanw/esbuild/issues/1343#issuecomment-854055814, https://github.com/microsoft/TypeScript/issues/16577#issuecomment-703190339, https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937

## Example

In [this example project derived from `example/simple`](https://github.com/mon-jai/ts-unused-exports/tree/master/example/simple), `inc()` from `math.ts` was imported with `.js` extension.

https://github.com/pzavolinsky/ts-unused-exports/blob/ad208c946855f9f5456d3d20435a4282f946c72b/example/simple/app.ts#L1

And TypeScript recognize it sucessfully.

![56-18-22-235633](https://user-images.githubusercontent.com/91261297/154717100-42f6acec-fa92-46ab-a83c-987cd85848a5.png)

However, ts-unused-exports failed to recognize that the default import of`math.ts` was imported in `app.ts`.

![Screenshot 2022-02-19 000527](https://user-images.githubusercontent.com/91261297/154718960-670bd156-6dc6-484b-a643-aba2c3825007.png)

Importing without `.js` extension won't work if the target of your `tsc` output is ES6 or above (e.g. developing an ESM npm package).

For example, in [this modified example/simple project](https://github.com/pzavolinsky/ts-unused-exports/commit/56939d1730404f57013d8e032ea9d5ba5b1b6f89), I am targeting ES6 for TypeScript output.

https://github.com/pzavolinsky/ts-unused-exports/blob/56939d1730404f57013d8e032ea9d5ba5b1b6f89/example/simple/tsconfig.json#L3-L4

Without having `.js` extension in import statements of your **TypeScript files**, the compiled output won't run, as [TypeScript won't add the extension for you](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937).

```
$ tsc && node app.js
node:internal/process/esm_loader:94
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'D:\GitHub\ts-unused-exports\example\simple\math' imported from D:\GitHub\ts-unused-exports\example\simple\app.js
Did you mean to import ../math.js?
    at new NodeError (node:internal/errors:371:5)
    at finalizeResolution (node:internal/modules/esm/resolve:394:11)
    at moduleResolve (node:internal/modules/esm/resolve:944:10)
    at defaultResolve (node:internal/modules/esm/resolve:1041:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v17.4.0
```

If TypeScript output target is ES6 or above (e.g. developing an ESM npm package) and **TypeScript source files** do not have `.js` extension in their import statements, the compiled output won't run on Node.js.

In fact, example/simple only works when the output format is set to commonjs. If I [modify it to output ESM instead](https://github.com/pzavolinsky/ts-unused-exports/commit/56939d1730404f57013d8e032ea9d5ba5b1b6f89),

https://github.com/pzavolinsky/ts-unused-exports/blob/56939d1730404f57013d8e032ea9d5ba5b1b6f89/example/simple/tsconfig.json#L3-L4

the compiled output won't run, as [TypeScript won't add the extension for you](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937).

```
$ tsc && node app.js
node:internal/process/esm_loader:94
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'D:\GitHub\ts-unused-exports\example\simple\math' imported from D:\GitHub\ts-unused-exports\example\simple\app.js
Did you mean to import ../math.js?
    at new NodeError (node:internal/errors:371:5)
    at finalizeResolution (node:internal/modules/esm/resolve:394:11)
    at moduleResolve (node:internal/modules/esm/resolve:944:10)
    at defaultResolve (node:internal/modules/esm/resolve:1041:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v17.4.0
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,8 @@
+
+### Added
+
+Allow for importing of 'x.js' files as opposed to simply 'x'. This aims towards supporting es6 and higher modules as opposed to CommonJS.
+
 ## [8.0.5] - 10 Dec 2022
 
 ### Changed

--- a/example/simple-es5-js/app.ts
+++ b/example/simple-es5-js/app.ts
@@ -1,0 +1,3 @@
+import inc from './math.js';
+
+console.log('two', inc(1));

--- a/example/simple-es5-js/execute.sh
+++ b/example/simple-es5-js/execute.sh
@@ -1,0 +1,1 @@
+npx tsc && node ./lib/app.js

--- a/example/simple-es5-js/math.ts
+++ b/example/simple-es5-js/math.ts
@@ -1,0 +1,6 @@
+export function unused1(x: number) { return x + 1; }
+
+// ts-unused-exports:disable-next-line
+export function unusedButDisabled(x: number) { return x + 2; }
+
+export default (x: number) => x + 1;

--- a/example/simple-es5-js/tsconfig.json
+++ b/example/simple-es5-js/tsconfig.json
@@ -1,0 +1,11 @@
+{
+    "compilerOptions":
+    {
+      "target": "ES5",
+      "module": "commonjs",
+      "noImplicitAny": true,
+      "strictNullChecks": true,
+      "outDir": "./lib",
+    },
+    "include": ["./**/*.ts"],
+  }

--- a/example/simple-es5-js/unused.ts
+++ b/example/simple-es5-js/unused.ts
@@ -1,0 +1,1 @@
+export function unused(x: number) { return x + 2; }

--- a/example/simple-es6/app.ts
+++ b/example/simple-es6/app.ts
@@ -1,0 +1,3 @@
+import inc from './math';
+
+console.log('two', inc(1));

--- a/example/simple-es6/app.ts
+++ b/example/simple-es6/app.ts
@@ -1,3 +1,3 @@
-import inc from './math';
+import * as inc from './math.js';
 
-console.log('two', inc(1));
+console.log('two', inc.default(1));

--- a/example/simple-es6/execute.sh
+++ b/example/simple-es6/execute.sh
@@ -1,0 +1,1 @@
+npx tsc && node ./lib/app.js

--- a/example/simple-es6/math.ts
+++ b/example/simple-es6/math.ts
@@ -1,0 +1,6 @@
+export function unused1(x: number) { return x + 1; }
+
+// ts-unused-exports:disable-next-line
+export function unusedButDisabled(x: number) { return x + 2; }
+
+export default (x: number) => x + 1;

--- a/example/simple-es6/math.ts
+++ b/example/simple-es6/math.ts
@@ -1,4 +1,4 @@
-export function unused1(x: number) { return x + 1; }
+export function unused1(x: number) { return x + 100; }
 
 // ts-unused-exports:disable-next-line
 export function unusedButDisabled(x: number) { return x + 2; }

--- a/example/simple-es6/package.json
+++ b/example/simple-es6/package.json
@@ -1,0 +1,15 @@
+{
+    "description": "Example used by integration tests (ispec)",
+    "repository": {
+        "type": "git",
+        "url": "https://github.com/pzavolinsky/ts-unused-exports.git"
+    },
+    "license": "MIT",
+    "scripts": {
+        "test": "npm i && ../../bin/ts-unused-exports ./tsconfig.json"
+    },
+    "type": "module",
+    "dependencies": {
+        "typescript": "^4.9.4"
+    }
+}

--- a/example/simple-es6/tsconfig.json
+++ b/example/simple-es6/tsconfig.json
@@ -3,7 +3,8 @@
     "target": "es6",
     "module": "es6",
     "noImplicitAny": true,
-    "strictNullChecks": true
+    "strictNullChecks": true,
+    "outDir": "./lib",
   },
   "include": ["./**/*.ts"],
 }

--- a/example/simple-es6/tsconfig.json
+++ b/example/simple-es6/tsconfig.json
@@ -1,0 +1,9 @@
+{
+  "compilerOptions": {
+    "target": "es6",
+    "module": "es6",
+    "noImplicitAny": true,
+    "strictNullChecks": true
+  },
+  "include": ["./**/*.ts"],
+}

--- a/example/simple-es6/unused.ts
+++ b/example/simple-es6/unused.ts
@@ -1,0 +1,1 @@
+export function unused(x: number) { return x + 2; }

--- a/example/simple/execute.sh
+++ b/example/simple/execute.sh
@@ -1,0 +1,1 @@
+npx tsc && node ./lib/app.js

--- a/example/simple/tsconfig.json
+++ b/example/simple/tsconfig.json
@@ -1,10 +1,11 @@
 {
-  "compilerOptions": {
+  "compilerOptions":
+  {
     "target": "ES5",
     "module": "commonjs",
-    "noEmit": true,
     "noImplicitAny": true,
-    "strictNullChecks": true
+    "strictNullChecks": true,
+    "outDir": "./lib",
   },
   "include": ["./**/*.ts"],
 }

--- a/features/import-alias.feature
+++ b/features/import-alias.feature
@@ -14,8 +14,8 @@ Scenario: Used export default - import * as x
   Given file "a.ts" is export default 1;
   And file "b.ts" is import * as ax from './a';
   When analyzing "tsconfig.json"
-  Then the result at a.ts is ["default"]
-# see star.feature - Import * marks 'default' as unused
+  Then the result is {}
+# see star.feature - Import * does NOT mark 'default' as unused
 
 Scenario: Used export default - import x
   Given file "a.ts" is export default 1;

--- a/features/star.feature
+++ b/features/star.feature
@@ -1,6 +1,6 @@
 Feature: import * / export *
 
-Scenario: Import * marks 'default' as unused
+Scenario: Import * does NOT mark 'default' as unused
   Given file "a.ts" is
     """
     export const a = 1;
@@ -8,8 +8,8 @@ Scenario: Import * marks 'default' as unused
     """
   And file "b.ts" is import * as all from './a';
   When analyzing "tsconfig.json"
-  Then the result at a.ts is ["default"]
-# note: default could still be used, as all.default
+  Then the result is {}
+# note: a or default could still be used, as all.a or all.default - so safest to mark as used
 
 Scenario: Import * and default
   Given file "a.ts" is

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "ts-unused-exports",
-  "version": "8.0.5",
+  "version": "9.0.0",
   "description": "ts-unused-exports finds unused exported symbols in your Typescript project",
   "main": "lib/app.js",
   "repository": {

--- a/src/analyzer.ts
+++ b/src/analyzer.ts
@@ -4,7 +4,11 @@ import {
   File,
   LocationInFile,
 } from './types';
-import { indexCandidates, removeExportStarPrefix } from './parser/util';
+import {
+  indexCandidates,
+  removeExportStarPrefix,
+  removeFileExtensionToAllowForJs,
+} from './parser/util';
 
 export { Analysis } from './types';
 
@@ -67,18 +71,9 @@ const getExportMap = (files: File[]): ExportMap => {
   return map;
 };
 
-const removeExtension = (filename: string): string => {
-  if (filename.endsWith('.js')) {
-    return filename.substring(0, filename.length - 3);
-  }
-  return filename;
-};
-
 const processImports = (file: File, exportMap: ExportMap): void => {
   Object.keys(file.imports).forEach((key) => {
-    const keyNoExtensionToAllowForJs = removeExtension(key);
-
-    let ex = exportMap[keyNoExtensionToAllowForJs]?.exports;
+    let ex = exportMap[removeFileExtensionToAllowForJs(key)]?.exports;
 
     // Handle imports from an index file
     if (!ex && key === '.') {

--- a/src/analyzer.ts
+++ b/src/analyzer.ts
@@ -67,9 +67,18 @@ const getExportMap = (files: File[]): ExportMap => {
   return map;
 };
 
+const removeExtension = (filename: string): string => {
+  if (filename.endsWith('.js')) {
+    return filename.substring(0, filename.length - 3);
+  }
+  return filename;
+};
+
 const processImports = (file: File, exportMap: ExportMap): void => {
   Object.keys(file.imports).forEach((key) => {
-    let ex = exportMap[key]?.exports;
+    const keyNoExtensionToAllowForJs = removeExtension(key);
+
+    let ex = exportMap[keyNoExtensionToAllowForJs]?.exports;
 
     // Handle imports from an index file
     if (!ex && key === '.') {
@@ -101,13 +110,9 @@ const processImports = (file: File, exportMap: ExportMap): void => {
       ex[imp].usageCount++;
     };
 
-    file.imports[key].forEach((imp) =>
-      imp === '*'
-        ? Object.keys(ex)
-            .filter((e) => e != 'default')
-            .forEach(addUsage)
-        : addUsage(imp),
-    );
+    file.imports[key].forEach((imp) => {
+      imp === '*' ? Object.keys(ex).forEach(addUsage) : addUsage(imp);
+    });
   });
 };
 

--- a/src/parser/import.ts
+++ b/src/parser/import.ts
@@ -12,15 +12,33 @@ import path = require('path');
 
 // Parse Imports
 
-const EXTENSIONS = ['.d.ts', '.ts', '.tsx', '.js', '.jsx'];
+const EXTENSIONS = [
+  '.d.ts',
+  '.ts',
+  '.cts',
+  '.mts',
+  '.tsx',
+  '.js',
+  '.cjs',
+  '.mjs',
+  '.jsx',
+];
 
 const isRelativeToBaseDir = (baseDir: string, from: string): boolean =>
   existsSync(resolve(baseDir, `${from}.js`)) ||
+  existsSync(resolve(baseDir, `${from}.cjs`)) ||
+  existsSync(resolve(baseDir, `${from}.mjs`)) ||
   existsSync(resolve(baseDir, `${from}.ts`)) ||
+  existsSync(resolve(baseDir, `${from}.cts`)) ||
+  existsSync(resolve(baseDir, `${from}.mts`)) ||
   existsSync(resolve(baseDir, `${from}.d.ts`)) ||
   existsSync(resolve(baseDir, `${from}.tsx`)) ||
   existsSync(resolve(baseDir, from, 'index.js')) ||
+  existsSync(resolve(baseDir, from, 'index.cjs')) ||
+  existsSync(resolve(baseDir, from, 'index.mjs')) ||
   existsSync(resolve(baseDir, from, 'index.ts')) ||
+  existsSync(resolve(baseDir, from, 'index.cts')) ||
+  existsSync(resolve(baseDir, from, 'index.mts')) ||
   existsSync(resolve(baseDir, from, 'index.tsx'));
 
 const joinWithBaseUrl = (baseUrl: string, from: string) => {

--- a/src/parser/util.ts
+++ b/src/parser/util.ts
@@ -9,13 +9,17 @@ export function isUnique<T>(value: T, index: number, self: T[]): boolean {
 export const indexCandidates = [
   '/index',
   '/index.ts',
+  '/index.cts',
+  '/index.mts',
   '/index.tsx',
   '/index.js',
+  '/index.cjs',
+  '/index.mjs',
 ];
 
 export function removeFileExtensionToAllowForJs(path: string): string {
   // ref: https://www.typescriptlang.org/docs/handbook/esm-node.html
-  const extensionsToStrip = ['.js', '.mjs', '.cjs'];
+  const extensionsToStrip = ['.js', '.cjs', '.mjs'];
 
   for (let i = 0; i < extensionsToStrip.length; i++) {
     const ext = extensionsToStrip[i];
@@ -26,8 +30,6 @@ export function removeFileExtensionToAllowForJs(path: string): string {
 }
 
 export function removeExportStarPrefix(path: string): string {
-  path = removeFileExtensionToAllowForJs(path);
-
   if (path.startsWith('*:')) return path.slice(2);
   else if (path.startsWith('*as:')) return path.slice(4);
 

--- a/src/parser/util.ts
+++ b/src/parser/util.ts
@@ -14,6 +14,8 @@ export const indexCandidates = [
 ];
 
 export function removeExportStarPrefix(path: string): string {
+  if (path.endsWith('.js')) path = path.substring(0, path.length - 3);
+
   if (path.startsWith('*:')) return path.slice(2);
   else if (path.startsWith('*as:')) return path.slice(4);
 

--- a/src/parser/util.ts
+++ b/src/parser/util.ts
@@ -13,8 +13,20 @@ export const indexCandidates = [
   '/index.js',
 ];
 
+export function removeFileExtensionToAllowForJs(path: string): string {
+  // ref: https://www.typescriptlang.org/docs/handbook/esm-node.html
+  const extensionsToStrip = ['.js', '.mjs', '.cjs'];
+
+  for (let i = 0; i < extensionsToStrip.length; i++) {
+    const ext = extensionsToStrip[i];
+    if (path.endsWith(ext)) return path.substring(0, path.length - ext.length);
+  }
+
+  return path;
+}
+
 export function removeExportStarPrefix(path: string): string {
-  if (path.endsWith('.js')) path = path.substring(0, path.length - 3);
+  path = removeFileExtensionToAllowForJs(path);
 
   if (path.startsWith('*:')) return path.slice(2);
   else if (path.startsWith('*as:')) return path.slice(4);

